File: Packages/OmniaApplication/Sources/OmniaApplication/ProviderValidationService.swift
import Foundation
import OmniaDomain

/// Candidate values required to test a provider form without persisting it.
public struct ProviderConnectionTestRequest: Sendable {
    public let provider: ProviderIdentity?
    public let endpoint: String
    public let model: String?
    public let credential: Credential?
    /// The API family the candidate connection targets; the family selects
    /// which Infrastructure inspector validates it (ARC-004, DES-011 §3.4).
    public let apiKind: ProviderAPIKind

    public init(
        provider: ProviderIdentity? = nil,
        endpoint: String,
        model: String? = nil,
        credential: Credential? = nil,
        apiKind: ProviderAPIKind = ProviderAPIKind.default
    ) {
        self.provider = provider
        self.endpoint = endpoint
        self.model = model
        self.credential = credential
        self.apiKind = apiKind
    }
}

public struct ProviderConnectionTestResult: Equatable, Sendable {
    public let models: [ModelReference]

    public init(models: [ModelReference]) {
        self.models = models
    }
}

/// Validates either an unsaved provider candidate or an existing connection
/// through injected transport paths selected by the connection's API family.
public struct ProviderValidationService: Sendable {
    private let testCandidate: @Sendable (
        URL,
        Credential,
        ModelReference?,
        ProviderAPIKind
    ) async throws -> [ModelReference]
    private let testExisting: @Sendable (
        ProviderIdentity,
        URL,
        ModelReference?,
        ProviderAPIKind
    ) async throws -> [ModelReference]

    public init(
        testCandidate: @escaping @Sendable (
            URL,
            Credential,
            ModelReference?,
            ProviderAPIKind
        ) async throws -> [ModelReference],
        testExisting: @escaping @Sendable (
            ProviderIdentity,
            URL,
            ModelReference?,
            ProviderAPIKind
        ) async throws -> [ModelReference]
    ) {
        self.testCandidate = testCandidate
        self.testExisting = testExisting
    }

    public func test(
        _ request: ProviderConnectionTestRequest
    ) async throws -> ProviderConnectionTestResult {
        let trimmedEndpoint = request.endpoint.trimmingCharacters(in: .whitespacesAndNewlines)
        guard
            let url = URL(string: trimmedEndpoint),
            let scheme = url.scheme?.lowercased(),
            scheme == "http" || scheme == "https",
            url.host != nil
        else {
            throw ProviderConnectionTestError.invalidEndpoint
        }
        let model = request.model?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let reference = model.flatMap { $0.isEmpty ? nil : ModelReference(name: $0) }
        let models: [ModelReference]
        if let provider = request.provider {
            models = try await testExisting(provider, url, reference, request.apiKind)
        } else {
            guard let credential = request.credential else {
                throw ProviderConnectionTestError.invalidCredential
            }
            models = try await testCandidate(url, credential, reference, request.apiKind)
        }
        return ProviderConnectionTestResult(models: models)
    }
}

File: Android/core/application/src/main/kotlin/com/omnia/application/ProviderValidationService.kt
package com.omnia.application

import com.omnia.domain.ModelReference
import com.omnia.domain.ProviderAPIKind
import com.omnia.domain.ProviderConnectionTestError
import com.omnia.domain.ProviderIdentity

class ProviderValidationService(
    private val testCandidate: suspend (endpoint: String, credential: com.omnia.domain.Credential, model: ModelReference?, apiKind: ProviderAPIKind) -> List<ModelReference>,
    private val testExisting: suspend (identity: ProviderIdentity, endpoint: String, model: ModelReference?, apiKind: ProviderAPIKind) -> List<ModelReference>,
) {
    @Throws(ApplicationValidationError::class)
    suspend fun test(request: ProviderConnectionTestRequest): ProviderConnectionTestResult {
        val endpoint = request.endpoint?.trim() ?: ""
        val model = request.model?.trim()?.takeIf { it.isNotEmpty() }?.let { ModelReference(it) }
        val apiKind = request.apiKind ?: ProviderAPIKind.default

        if (!isValidEndpoint(endpoint)) {
            throw ApplicationValidationError.Invalid("Invalid endpoint: $endpoint")
        }

        val credential = request.credential

        val models = if (request.providerIdentity != null) {
            testExisting(request.providerIdentity, endpoint, model, apiKind)
        } else {
            val resolvedCredential = credential
                ?: throw ApplicationValidationError.Invalid("Credential is required")
            testCandidate(endpoint, resolvedCredential, model, apiKind)
        }

        return ProviderConnectionTestResult(models = models)
    }

    private fun isValidEndpoint(endpoint: String): Boolean {
        if (endpoint.isBlank()) return false
        val lowered = endpoint.lowercase()
        return lowered.startsWith("http://") || lowered.startsWith("https://")
    }
}

data class ProviderConnectionTestRequest(
    val providerIdentity: ProviderIdentity? = null,
    val endpoint: String? = null,
    val model: String? = null,
    val credential: com.omnia.domain.Credential? = null,
    val apiKind: ProviderAPIKind? = null,
)

data class ProviderConnectionTestResult(
    val models: List<ModelReference>,
)
